Content.Server/Decals/DecalSystem.cs

```diff
@@ -0,0 +1,323 @@
+        }
+
+        private void OnTileChanged(object? sender, TileChangedEventArgs e)
+        {
+            if (!e.NewTile.IsSpace())
+                return;
+
+            var chunkCollection = ChunkCollection(e.NewTile.GridIndex);
+            var indices = GetChunkIndices(e.NewTile.GridIndices);
+            var toDelete = new HashSet<uint>();
+            if (chunkCollection.TryGetValue(indices, out var chunk))
+            {
+                foreach (var (uid, decal) in chunk)
+                {
+                    if (new Vector2((int) Math.Floor(decal.Coordinates.X), (int) Math.Floor(decal.Coordinates.Y)) ==
+                        e.NewTile.GridIndices)
+                    {
+                        toDelete.Add(uid);
+                    }
+                }
+            }
+
+            if (toDelete.Count == 0) return;
+
+            foreach (var uid in toDelete)
+            {
+                RemoveDecalInternal(e.NewTile.GridIndex, uid);
+            }
+
+            DirtyChunk(e.NewTile.GridIndex, indices);
+        }
+
+        private void OnPlayerStatusChanged(object? sender, SessionStatusEventArgs e)
+        {
+            switch (e.NewStatus)
+            {
+                case SessionStatus.InGame:
+                    _previousSentChunks[e.Session] = new();
+                    break;
+                case SessionStatus.Disconnected:
+                    _previousSentChunks.Remove(e.Session);
+                    break;
+            }
+        }
+
+        protected override void DirtyChunk(GridId id, Vector2i chunkIndices)
+        {
+            if(!_dirtyChunks.ContainsKey(id))
+                _dirtyChunks[id] = new HashSet<Vector2i>();
+            _dirtyChunks[id].Add(chunkIndices);
+        }
+
+        public bool TryAddDecal(string id, EntityCoordinates coordinates, [NotNullWhen(true)] out uint? uid, Color? color = null, Angle? rotation = null, int zIndex = 0, bool cleanable = false)
+        {
+            uid = 0;
+            if (!PrototypeManager.HasIndex<DecalPrototype>(id))
+                return false;
+
+            var gridId = coordinates.GetGridId(EntityManager);
+            if (MapManager.GetGrid(gridId).GetTileRef(coordinates).IsSpace())
+                return false;
+
+            rotation ??= Angle.Zero;
+            var decal = new Decal(coordinates.Position, id, color, rotation.Value, zIndex, cleanable);
+            var chunkCollection = DecalGridChunkCollection(gridId);
+            uid = chunkCollection.NextUid++;
+            var chunkIndices = GetChunkIndices(decal.Coordinates);
+            if(!chunkCollection.ChunkCollection.ContainsKey(chunkIndices))
+                chunkCollection.ChunkCollection[chunkIndices] = new();
+            chunkCollection.ChunkCollection[chunkIndices][uid.Value] = decal;
+            ChunkIndex[gridId][uid.Value] = chunkIndices;
+            DirtyChunk(gridId, chunkIndices);
+            return true;
```

Comment:
holy wall of text man

Reply:
we do a little chunkchunkchunkchunkuiddecal